src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -1595,25 +1604,47 @@ internal static void CheckActionPreference(FunctionContext funcContext, Exceptio
             // set $? to false indicating an error
             context.QuestionMarkVariableValue = false;
 
-            bool anyTrapHandlers = funcContext._traps.Any() && funcContext._traps.Last().Item2 != null;
+            ActionPreference preference = GetErrorActionPreference(context);
 
-            if (!anyTrapHandlers && !ExceptionHandlingOps.NeedToQueryForActionPreference(rte, context))
+            // If the exception was not rethrown and we are not currently
+            // handling an exception, then the exception is new, and we
+            // can break on it if requested.
+            if (!rte.WasRethrown &&
+                context.CurrentExceptionBeingHandled == null &&
+                preference == ActionPreference.Break)
             {
-                throw rte;
+                context.Debugger?.Break(rte);
+
+                // set the value of $? here in case it was reset in the debugger.
+                context.QuestionMarkVariableValue = false;
             }
 
-            ActionPreference preference;
+            // Item2 in the trap tuples is the action (script) for the trap.
+            // A null action script is only used to indicate when exceptions
+            // should be thrown up to a higher level, and doesn't count as an
+            // actual trap handler in the function context.
+            bool anyTrapHandlers = funcContext._traps.Count > 0 && funcContext._traps[funcContext._traps.Count - 1].Item2 != null;
+
             if (anyTrapHandlers)
             {
+                // update the action preference according to how the exception is
+                // handled in the trap statement(s).
                 preference = ProcessTraps(funcContext, rte);
+
+                // set the value of $? here in case it was reset in trap handling.
+                context.QuestionMarkVariableValue = false;
             }
-            else
+            else if (ExceptionCannotBeStoppedContinuedOrIgnored(rte, context))
             {
-                preference = ExceptionHandlingOps.QueryForAction(rte, rte.Message, context);
+                throw rte;
```

Comment:
will `$?` get set as a result of this throw, or does that need to be done explicitly.

Reply:
It's explicitly set on line 1605, just a little above this.